File: cp/LC1372-Longest-ZigZag-Path-in-a-Binary-Tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
      
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        self.pathLength = 0
        
        def dfs(node, goLeft, steps):
            if node:
                self.pathLength = max(self.pathLength, steps)
                if goLeft:
                    dfs(node.left, False, steps + 1)
                    dfs(node.right, True, 1)
                else:
                    dfs(node.left, False, 1)
                    dfs(node.right, True, steps + 1)
        
        dfs(root, False, 0)
        dfs(root, True, 0)
        return self.pathLength
    
tr=TreeNode(1)
tr.right=TreeNode(2)
tr.right.left=TreeNode(3)
tr.right.left.right=TreeNode(4)
tr.right.left.right.left=TreeNode(5)

sol=Solution()

print(sol.longestZigZag(tr))

File: cp/LC1372-Longest-ZigZag-Path-in-a-Binary-Tree.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define mod 1000000007

class TreeNode
{
public:
    int data;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : data(0), left(nullptr), right(nullptr) {}
    TreeNode(int val)
    {
        data = val;
        left = NULL;
        right = NULL;
    }
    TreeNode(int x, TreeNode *left, TreeNode *right) : data(x), left(left), right(right) {}
};

class Solution
{
    void solve(TreeNode *root, int dir, int cnt, int &ans)
    {
        if (root == nullptr)
        {
            return;
        }
        if (dir == 0)
        {
            if (root->left != nullptr)
                solve(root->left, 0, 1, ans);
            if (root->right != nullptr)
                solve(root->right, 1, cnt + 1, ans);
        }
        else
        {
            if (root->left != nullptr)
                solve(root->left, 0, cnt + 1, ans);
            if (root->right != nullptr)
                solve(root->right, 1, 1, ans);
        }

        ans = max(ans, cnt);
        return;
    }

public:
    int longestZigZag(TreeNode *root)
    {
        int ans = 0;
        solve(root, -1, 0, ans);
        return ans;
    }
};

int main()
{
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->left->right->left = new TreeNode(8);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(7);
    root->right->right->left = new TreeNode(9);
    root->right->right->right = new TreeNode(10);
    Solution sol;
    cout << sol.longestZigZag(root) << endl;

    return 0;
}